libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance.hs

```diff
@@ -418,3 +504,140 @@ instance
+    pp = nes ^. nesEsL . curPParamsEpochStateL
+    txfee = txb ^. feeTxBodyL
+    s = nes ^. nesEsL . esLStateL . lsCertStateL
+    txsize = tx ^. sizeTxF
+    maxTxSize = pp ^. ppMaxTxSizeL
+    txouts = txb ^. outputsTxBodyL
+    maxValSize = pp ^. ppMaxValSizeL
+    serSize = fromIntegral . BSL.length . serialize (pvMajor (pp ^. ppProtocolVersionL))
+    consumedVal = consumed pp s utxo txb
+    producedVal = produced pp s txb
+  slot <- gets $ view impLastTickL
+  modifyNES $ const nes
+  agdaUtxoState <- case toSpecRep $ nes ^. nesEsL . esLStateL . lsUTxOStateL of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  agdaTx <- case toSpecRep tx of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  logEntry $ "Impl Tx:\n" <> showExpr tx
+  logEntry $ "Agda Tx:\n" <> showExpr agdaTx
+  logCurPParams
+  logEntry $
+    unlines
+      [ "txins ≢ ∅\n  " <> show (not $ null txins)
+      , " txins ⊆ dom utxo\n  " <> show (txins `isSubsetOf` Map.keysSet (unUTxO utxo))
+      , " inInterval slot txvldt\n  " <> show (inInterval slot txvldt)
+      , " minfee pp tx ≤ txfee\n  " <> show (getMinFeeTx pp tx <= txfee)
+      , " consumed pp s txb ≡ produced pp s txb\n  " <> show (consumedVal == producedVal)
+      , " txsize ≤ maxTxSize pp\n  " <> show (txsize <= fromIntegral maxTxSize)
+      , " ∀[ (_ , txout) ∈ txouts .proj₁ ] inject (utxoEntrySize txout * minUTxOValue pp) ≤ᵗ getValue txout\n  "
+          <> show (all (\txout -> inject (Coin 0) <= txout ^. valueTxOutL) txouts)
+      , " ∀[ (_ , txout) ∈ txouts .proj₁ ] serSize (getValue txout) ≤ maxValSize pp\n  "
+          <> show (all (\txout -> serSize (txout ^. valueTxOutL) <= maxValSize) txouts)
+      , " ∀[ (a , _) ∈ range txouts ] Sum.All (const ⊤) (λ a → a .BootstrapAddr.attrsSize ≤ 64) a\n  "
+          -- I assume we never create bootstrap addresses
+          <> show (all (\txout -> case txout ^. addrTxOutL of AddrBootstrap _ -> False; _ -> True) txouts)
+      ]
+  logEntry $
+    unlines
+      [ "Implementation:"
+      , "  consumed:\t" ++ show (coin consumedVal)
+      , "    ins:\t\t" ++ show (coinBalance (UTxO (Map.restrictKeys (unUTxO utxo) (txb ^. inputsTxBodyL))))
+      , "  produced:\t" ++ show (coin producedVal)
+      , "    outs:\t" ++ show (coin $ sumAllValue txouts)
+      , "    fee:\t\t" ++ show txfee
+      ]
+  lastTick <- gets $ view impLastTickL
+  pParams <- getsNES $ nesEsL . curPParamsEpochStateL
+  certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
+  let
+    utxoEnv =
+      UtxoEnv
+        { ueSlot = lastTick
+        , uePParams = pParams
+        , ueGenDelegs = certState ^. certDStateL . dsGenDelegsL
+        , ueCertState = certState
+        }
+  agdaUtxoEnv <- case toSpecRep utxoEnv of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  let
+    newAgdaUtxoState = Agda.utxoStep agdaUtxoEnv agdaUtxoState agdaTx
+  modify $ impDoFixupL .~ False
+  submitRes <- trySubmitTx tx
+  modify $ impDoFixupL .~ doFixup
```

Comment:
I thought it'd be clearer if it was something named rather than just some `Bool` argument to a function. Also you'll usually want to disable fixups for the whole test, not single transactions.

Reply:
This "Imp"erative approach is stating to affect your coding style.

You chose to go with a mutable variable over using a functional approach!

All you need to do is create a separate function that does fixup and another that does not. This way it is not a flag to a function, but two separate functions with a separate name which can be used in all those tests that do or do not need fixing up.

Reply:
Ok, so I tried the approach you suggested, but I feel that I'd have to introduce so many functions to do the thing that I want to do. For example I'd have to make `trySubmitTxConform` and `trySubmitTxConformNoFixup` instead of just toggling a single boolean in the state.

Reply:
Maybe the better solution would be to introduce a `withNoFixup` combinator that would toggle the flag to `True` for a single command. It's a pretty widespread pattern from what I've seen in other libraries.

Reply:
In fact, I think we should add `withExpectingFailure` instead of having all the `submitTxFailing`/`submitTx` kind of functions.

Reply:
We could also do that via a `Reader` monad I think by using the `local` function

Reply:
If you can solve this problem with a `Reader`, you can solve it with a simple function argument (that is what ReaderT is afterall), which is what I am suggesting you do

Reply:
I still think it's better to write `withNoFixup . withExpectFailure . submitTx` than having a separate function for every combination of configurations or forcing the user to pass a bunch of `Bool`s every time they try to submit a transaction.

Reply:
This pattern is used in [QuickCheck](https://hackage.haskell.org/package/QuickCheck-2.14.3/docs/src/Test.QuickCheck.Gen.html#resize) and [hspec](https://hackage.haskell.org/package/hspec-2.11.7/docs/Test-Hspec.html#v:before_) also

Reply:
It is very different. What you are doing is modifying a mutable variable, where the approaches you mentioned above are using modifying functions of the immutable scope variables, similar to `local` on Reader.

The thing you are missing is the presence of exceptions! Which makes your approach totally invalid.

The easiest way to solve it is with what I suggested: using an argument to a function. Or if you want to add extra immutable environment to the ImpTestM by the means of ReaderT, be my guest. But the solution you have right now is not gonna work

Reply:
I went with the `local` approach